[HomePage] Add recent work section and additional styling for home page

abbyhu2000 opened the pull request:
### Description

This PR implements: 

1. Re-styling of the home page when advanced setting`home:useNewHomePage` is toggled on. Mocks are here: #6066
2. Add a recent work section with max 8 cards; any dashboard, visualization, saved search that are accessed will be added to the recent work. The most recent one will show up as the first. Each recent work card has work item name, type and last updated time.
3. If observability plugin is installed, it will show an extra card for observability in `Start working with data` section

### Issues Resolved

partially resolves #6066

## Screenshot

1. Home page toggle
https://github.com/opensearch-project/OpenSearch-Dashboards/assets/43937633/7b700395-e1cd-4572-b81c-7f24b5aee8fd

2. New home page without observability plugin
<img width="866" alt="Screenshot 2024-07-08 at 3 01 14 PM" src="https://github.com/opensearch-project/OpenSearch-Dashboards/assets/43937633/20e61336-0ef8-4376-bceb-8837ad7d9641">


3. New home page with observability plugin
<img width="866" alt="Screenshot 2024-07-08 at 3 05 29 PM" src="https://github.com/opensearch-project/OpenSearch-Dashboards/assets/43937633/204ced77-e6d8-47ff-af3b-fed4b907fbff">


4. Section styling and links
https://github.com/opensearch-project/OpenSearch-Dashboards/assets/43937633/821d8aa4-e988-4adf-9556-2e05a76eef08


5. Recent work section
https://github.com/opensearch-project/OpenSearch-Dashboards/assets/43937633/7aeaa20e-8d0a-4284-b5f7-c55dde949d4a


## Changelog
- feat: Add recent work section and additional styling for home page

ashwin-pc reviewed (commented):
Two things i want to call out here:
1. How do i see the recently accessed section? Pulling down the PR and loading the new homepage does not have this section. Also can you add a screenshot to the PR so that people who dont pull down the change can see it too?
2. You are modifying the architecture of the section component. That seems unnecessary for the filter but want to hear more on why you think the option you proposed is the best way to do it and what are the other alternitives you've tried.

abbyhu2000 commented on src/plugins/discover/public/application/view_components/utils/use_search.ts:
```diff
@@ -290,7 +290,8 @@ export const useSearch = (services: DiscoverViewServices) => {
           savedSearchInstance.title,
-          savedSearchInstance.id
+          savedSearchInstance.id,
+          savedSearchInstance.getOpenSearchType()
```
I am adding it to use for the timestamp on the recent work section card, its WIP.

<img width="781" alt="Screenshot 2024-03-28 at 3 31 04 PM" src="https://github.com/opensearch-project/OpenSearch-Dashboards/assets/43937633/46b18713-4578-4ce3-afaa-3ad1a111a507">

abbyhu2000 replied:
Finished adding the `last updated by` in the latest commit.

abbyhu2000 commented on src/plugins/home/public/application/components/homepage/section.tsx:
```diff
@@ -65,28 +36,7 @@ export const Section: FC<Props> = ({ render, title, description, links }) => {
       color="transparent"
       data-test-subj="homepageSection"
     >
-      <EuiFlexGroup direction="row" alignItems="center" gutterSize="s" responsive={false}>
```
The other reason i re-construct here is also because different sections now have a slightly different header: `recentWork` section has a filter button next to the toggle, `learn basics` section has a link next to the toggle. 

If we do not want to change the structure here at all, i guess we can add reactNode as another opt parameter for section, so they can choose to render that next to the toggle if they pass in a reactNode.

ashwin-pc replied:
Yep let's do that!

abbyhu2000 commented on src/plugins/home/public/application/components/homepage/footer.tsx:
```diff
@@ -68,38 +68,6 @@ export const Footer: React.FC = () => {
     <EuiFlexGroup direction="row" wrap>
       {isAdvancedSettingsEnabled && <EuiFlexItem grow={false}>{defaultRouteButton}</EuiFlexItem>}
-
-      <EuiFlexItem grow={false}>
-        <RedirectAppLinks application={application}>
-          <EuiButtonEmpty
-            flush="both"
-            href={getUrlForApp('home', { path: '#/feature_directory' })}
-            iconType="apps"
-            size="xs"
-          >
-            <FormattedMessage
-              id="home.footer.appDirectoryButtonLabel"
-              defaultMessage="View app directory"
-            />
-          </EuiButtonEmpty>
-        </RedirectAppLinks>
-      </EuiFlexItem>
-
-      <EuiFlexItem grow={false}>
-        <RedirectAppLinks application={application}>
-          <EuiButtonEmpty
-            flush="both"
-            href={getUrlForApp('opensearch_dashboards_overview')}
-            iconType="visualizeApp"
-            size="xs"
-          >
-            <FormattedMessage
-              id="home.footer.visualizeAndAnalyze"
-              defaultMessage="Visualize & Analyze"
-            />
-          </EuiButtonEmpty>
-        </RedirectAppLinks>
-      </EuiFlexItem>
```
yes based on the latest UI mock

abbyhu2000 commented:
Added a comment here about the necessity of adding a filter button on the recent work section: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/6066#issuecomment-2031004270

Changed the rendering structure of section back to previous way.

SuZhou-Joe commented on src/plugins/discover/public/application/view_components/utils/use_search.ts:
```diff
@@ -290,7 +291,9 @@ export const useSearch = (services: DiscoverViewServices) => {
-          savedSearchInstance.id
+          savedSearchInstance.id,
+          savedSearchInstance.getOpenSearchType(),
+          moment(Date.now()).format('MM/DD/YYYY HH:mm')
```
Can we use timestamp(number) here? It is more convenient to do the sort by using that. And date format should comply with user's localization setting and can be done when render the timestamp.

SuZhou-Joe commented on src/core/public/chrome/recently_accessed/recently_accessed_service.ts:
```diff
@@ -56,11 +58,13 @@ export class RecentlyAccessedService {
     return {
       /** Adds a new item to the history. */
-      add: (link: string, label: string, id: string) => {
+      add: (link: string, label: string, id: string, type?: string, updatedAt?: string) => {
```
```suggestion
      add: (link: string, label: string, id: string, extraProps?: { type?: string; updatedAt?: string }) => {
```
What about we using a object? Or the arguments list may be too long to use.

abbyhu2000 replied:
Changed

SuZhou-Joe commented on src/plugins/dashboard/public/application/utils/use/use_saved_dashboard_instance.ts:
```diff
@@ -129,7 +130,9 @@ export const useSavedDashboardInstance = ({
-              dashboardIdFromUrl
+              dashboardIdFromUrl,
+              savedDashboard.getOpenSearchType(),
+              moment(Date.now()).format('MM/DD/YYYY HH:mm')
```
```suggestion
              Date.now()
```
Can we just pass `Date.now()` to `add` method? I think timestamp is more generic than a formatted string.

abbyhu2000 replied:
changed to Date.now()

SuZhou-Joe commented on src/plugins/home/public/application/components/homepage/sections/recent_work.tsx:
```diff
@@ -0,0 +1,94 @@
+                    />
+                    {recentWorkItem[0].name}
+                    <br />
+                    {'Last updated ' + moment(recentAccessItem?.updatedAt).fromNow()}
```
should be wrapped by i18n maybe?

abbyhu2000 replied:
wrapped label, name and updatedAt in i18n

BionIT commented:
Hi @abbyhu2000 , is this PR targeting 2.15 or 2.16?

SuZhou-Joe commented on src/core/public/chrome/recently_accessed/recently_accessed_service.ts:
```diff
@@ -59,15 +61,32 @@ export class RecentlyAccessedService {
-          id,
-          ...(currentWorkspaceId && { workspaceId: currentWorkspaceId }),
-        });
+        if (extraProps) {
+          const type = extraProps!.type;
+          const updatedAt = extraProps!.updatedAt;
+          history.add({
+            link,
+            label,
+            id,
+            ...(currentWorkspaceId && { workspaceId: currentWorkspaceId }),
+            type,
+            updatedAt,
+          });
+        } else {
+          history.add({
+            link,
+            label,
+            id,
+            ...(currentWorkspaceId && { workspaceId: currentWorkspaceId }),
+          });
+        }
```
```suggestion
      history.add({
        link,
        label,
        id,
        ...(currentWorkspaceId && { workspaceId: currentWorkspaceId }),
        ...extraProps,
      });
```
What about we simplifying the code like this?

SuZhou-Joe commented on src/plugins/home/public/application/components/homepage/sections/recent_work.tsx:
```diff
@@ -0,0 +1,103 @@
+const render = renderFn(() => {
+  const services = getServices();
+  const navigateToUrl = services.application.navigateToUrl;
+  const recentAccessed = useObservable(services.chrome.recentlyAccessed.get$(), []);
```
I suppose the correct order here should be:
1. Filter by type
2. Slice to no more than 8 items
3. render the items

SuZhou-Joe commented on src/plugins/saved_objects/public/saved_object/helpers/save_saved_object.ts:
```diff
@@ -123,7 +125,9 @@ export async function saveSavedObject(
         savedObject.title,
-        String(savedObject.id)
+        String(savedObject.id),
+        savedObject.getOpenSearchType(),
+        moment(Date.now()).format('MM/DD/YYYY HH:mm')
```
```suggestion
        { type: savedObject.getOpenSearchType(), updateAt: Date.now() }
```
Should be like this?

SuZhou-Joe commented on src/plugins/vis_builder/public/application/utils/use/use_saved_vis_builder_vis.ts:
```diff
@@ -74,6 +74,13 @@ export const useSavedVisBuilderVis = (visualizationIdFromUrl: string | undefined
           data.query.filterManager.setAppFilters(actualFilters);
           data.query.queryString.setQuery(query);
 
+          chrome.recentlyAccessed.add(
```
Seems the call does not comply with the interface of `chrome.recentlyAccessed.add`.

ashwin-pc reviewed (commented):
Did a cursory review. Have some questions about the implementation details but given that we are going a different direction with the homepage, do we still need this PR?

SuZhou-Joe commented:
@ashwin-pc At least the recent items section is still needed.

ananzh commented:
@abbyhu2000 is this still 2.16?

abbyhu2000 commented:
This doesn't target 2.16 anymore. Will close
